Add documentation for Wait.On and WaitOn transforms

dizzydroid opened the pull request:
This PR implements the documentation for the `Wait.On` (Java) and `WaitOn` (Python) transforms in Apache Beam. The updates aim to provide clearer guidance and more comprehensive examples for users.

### Changes
1. **Java `Wait.On`**:
   - Added examples demonstrating usage with multiple signals and streaming mode with windowing.
   - Added reference in `transforms/java/overview` page

2. **Python `WaitOn`**:
   - Created a detailed documentation file with examples for basic usage, multiple signals, and streaming mode.
   - Added reference in `transforms/python/overview` page

### Purpose
These changes improve the usability of the documentation by providing practical examples and highlighting related transforms that can be used in conjunction with `Wait.On` and `WaitOn`.

Resolves #35646

dizzydroid commented:
Thank you! In the basic example provided:
```python
# Example 1: Basic usage
with beam.Pipeline() as p:
    main = p | 'CreateMain' >> beam.Create([1, 2, 3])
    side = p | 'CreateSide' >> beam.Create(['a', 'b', 'c'])

    result = main | 'WaitOnSide' >> WaitOn(side)
    result | beam.Map(print)
 ```

the line `result = main | 'WaitOnSide' >> WaitOn(side)` implements a coordination barrier where `main` elements `[1, 2, 3]` pass through unchanged, but only after `side` processing completes. 

Internally, WaitOn converts `side` into a side input using `beam.Map(lambda x, *unused_sides: x, *sides)` - the main data (`x`) flows through unmodified while the side inputs (`*unused_sides`) act as completion signals. When `result | beam.Map(print)` executes, it prints `1, 2, 3` but with guaranteed ordering that `side` finished first.

So it's meant to be illustrating a synchronization primitive for coordinating pipeline stages, as commonly used for database write ordering or ensuring setup tasks complete before main processing.

I've added [explanatory comments](https://github.com/apache/beam/pull/35647/commits/3df0ac68188a72db5609f677a46f16ca67f34f90) to both the java and python examples, too. I hope this clears things up a little!

tvalentyn commented on website/www/site/content/en/documentation/transforms/python/other/waiton.md:
```diff
@@ -0,0 +1,65 @@
+# Example 1: Basic usage
+with beam.Pipeline() as p:
+    main = p | 'CreateMain' >> beam.Create([1, 2, 3])
+    side = p | 'CreateSide' >> beam.Create(['a', 'b', 'c'])
```
How about we change 'side' to 'signal' and add some processsing step after create, so that there is a processing step involved? Even if something simple as to sleep for a few seconds.

tvalentyn commented on website/www/site/content/en/documentation/transforms/python/other/waiton.md:
```diff
@@ -0,0 +1,65 @@
+
+# WaitOn
+
+`WaitOn` delays the processing of a main `PCollection` until one or more other `PCollections` (signals) have finished processing. This is useful for enforcing ordering or dependencies between different parts of a pipeline, especially when some outputs interact with external systems (such as writing to a database).
```
How about the following wording in both Python and Java docs:

`WaitOn` returns a  `PCollection` with the contents identical to the input `PCollection`, but  delays the downstream processing  until one or more other `PCollections` (signals) have finished processing. This is useful for enforcing ordering or dependencies between different parts of a pipeline, especially when some outputs interact with external systems (such as writing to a database).`

When you apply `WaitOn`, the elements of the main `PCollection` will not be emitted for downstream processing  until the computations required to produce the specified signal `PCollections` have completed. In streaming mode, this is enforced per window: the corresponding window of each waited-on `PCollection` must close before elements are passed through.

dizzydroid replied:
Fixed in [4370cc4](https://github.com/apache/beam/pull/35647/commits/4370cc4b3b2e00a73eac6e3e447eab1c6be02e71) and [3997502](https://github.com/apache/beam/pull/35647/commits/39975023bc2150646c7c8f98a109a32d0b218b34)

tvalentyn commented on website/www/site/content/en/documentation/transforms/python/other/waiton.md:
```diff
@@ -0,0 +1,65 @@
+with beam.Pipeline() as p:
+    main = (p | 'CreateMain' >> beam.Create([1, 2, 3])
+             | "ApplyWindow" >> beam.WindowInto(FixedWindows(5 * 60)))
+    side = (p | 'CreateSide' >> beam.Create(['a', 'b', 'c'])
```
There are not timestamps associated with elements after Create, and there will be an error if there is two transforms labelled 'ApplyWindow'. I would remove example 3

dizzydroid replied:
Fixed in [4370cc4](https://github.com/apache/beam/pull/35647/commits/4370cc4b3b2e00a73eac6e3e447eab1c6be02e71)

tvalentyn commented on website/www/site/content/en/documentation/transforms/java/other/wait.md:
```diff
@@ -0,0 +1,56 @@
+
+// Wait for 'signal' to complete before processing 'main'
+// Elements pass through unchanged after 'signal' finishes
+PCollection<String> result = main.apply(Wait.on(signal));
```
1. Could you please add a link to the javadoc: https://beam.apache.org/releases/javadoc/current/org/apache/beam/sdk/transforms/Wait.html

2. There is no downstream processing after  Wait.on - could we add something? You could also reuse the example from the Javadoc if you don't feel like making a full example.

dizzydroid replied:
Fixed in [3997502](https://github.com/apache/beam/pull/35647/commits/39975023bc2150646c7c8f98a109a32d0b218b34)

tvalentyn commented on website/www/site/content/en/documentation/transforms/python/overview.md:
```diff
@@ -82,4 +82,6 @@ limitations under the License.
   most useful for adjusting parallelism or preventing coupled failures.</td></tr>
   <tr><td><a href="/documentation/transforms/python/other/windowinto">WindowInto</a></td><td>Logically divides up or groups the elements of a collection into finite
   windows according to a function.</td></tr>
+  <tr><td><a href="/documentation/transforms/python/other/waiton">WaitOn</a></td><td>Delays processing of a PCollection until other PCollections have finished processing.</td></tr>
```
Could you please add Wait/WaitOn to the sidebard on the left hand side of the page? Thanks!
See (e.g.) Transform Catalog > Java > Other. Currently rendered version (available at https://apache-beam-website-pull-requests.storage.googleapis.com/35647/documentation/transforms/java/other/wait/index.html ) doesn't have it.

tvalentyn reviewed (changes requested):
Thanks!

tvalentyn commented:
Thanks! Could you please check that the examples can run in the Beam Playground?

I tried one Java sample in https://play.beam.apache.org/?sdk=java&shared=ZmuddB_rIbW and got an error (`error: incompatible types`).

Here is a possible fix: https://play.beam.apache.org/?sdk=java&shared=CfR_fav1XiA

Please check other examples too. Thanks!

dizzydroid commented:
Thanks for the update! I've added the required modifications to the first example and added logging to the second one, I've tested all the examples and they should be working now:
Second Java Example: https://play.beam.apache.org/?sdk=java&shared=tRniGwr4Rs7
Python Example 1: https://play.beam.apache.org/?sdk=python&shared=arrtgNbD9Vg
Python Example 2:  https://play.beam.apache.org/?sdk=python&shared=FEtHRsej4uv

dizzydroid commented:
<img width="1325" height="1798" alt="image" src="https://github.com/user-attachments/assets/2ebba483-4ced-454c-9c16-bdb1f31ce800" />

@tvalentyn , I have updated the code examples and they should now be rendered similar to other examples, let me know if this works!

dizzydroid commented on website/www/site/content/en/documentation/transforms/python/other/waiton.md:
```diff
@@ -30,27 +30,33 @@ import apache_beam as beam
-                | 'ProcessSignal1' >> beam.Map(lambda x: time.sleep(1) or x.upper()))
-    signal2 = (p | 'CreateSignal2' >> beam.Create(['x', 'y', 'z'])
-                | 'ProcessSignal2' >> beam.Map(lambda x: time.sleep(3) or x * 2))
+  with beam.Pipeline(options=pipeline_options) as p:
```
Nope, editor must have messed up. I have fixed this in 8fb2927, 9683238
